articles/express/index.md

```diff
@@ -0,0 +1,318 @@
+
+[Regular Expressions (RE)](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions) are patterns used to match character combinations in strings. We can use RE to match parameters and define our routes. For example in the example above if we wanted the productId to be only an integer we can try the following:
+
+```JavaScript
+...
+app.get(/^\/products\/(\d+)$/, function(req, res) {
+    var productId = parseInt(req.params[0], 10);
+    ...
+});
+...
+```
+
+### Template Engines
+Websites are built with HTML, you can dynamically generate HTML pages using Express. Dynamically generated HTML pages are useful when you want to show real time data or change a page's details based on the user. A template engine allows you to use static template files and at runtime replace variables in a template file with actual data. There are different template engines available like [Pug](https://pugjs.org/), Jade, and [EJS](https://ejs.co/). Lets see a basic template using EJS.
+
+First let's install it using npm. `npm install ejs` and create a directory called `views` to store your templates and HTML files.
+
+```JavaScript
+//app.js
+var express = require("express");
+
+app.set("view engine", "ejs"); //set view engine to ejs
+app.set("views", "views");     //set views directory
+
+app.get("/", function(req, res){   //res.render() renders a view and send the HTML to the client
+        res.render("index", {
+                message: "Hello and Welcome !!!" // this can be any thing you want
+        });
+});
+
+app.listen(3000);
+```
+
+Create a file called `index.ejs` in the views directory.
+
+```html
+<html>
+    <head>
+        <meta charset="utf-8">    
+    </head>
+
+    <body>
+        <%= message %>
+    </body>
+</html>
+```
+If you go to the webpage you can see `Hello and Welcome !!!`. You can write JavaScript expressions inside the `<%= exp %>`. Look at the [docs](https://ejs.co/#docs) for the complete syntax and rules.
+
+**Testing Express Applications**.
+
+Testing is an important part of developing software. Read [this post](/engineering-education/node-testing/) where we discuss testing Node.js applications using Mocha and Chai.
+
+### Conclusion
+The minimalistic philosophy of Express may not be suited for everyone's needs, because you can make mistakes while you are making those decisions about your applications infrastructure. One of the best practice is to have a directory structure like this or something similar to this for your express app
+
+```bash
+app/
+├── public/
+│   ├── bootstrap.css
+|   ├── favicon.ico
+├── views/
+│   ├── hello.ejs
+│   └── index.html
+└── routes/
+|   ├── api.js
+|   └── router.js
+├── app.js
+├── package.json
+├── tests/
+└── bin/
+```
+
+Express is an unopinionated framework which works best by making use of third party software to create a full-fledged application. Simplicity is a high level goal for software in general, when writing Express applications keep this in mind.
```

Comment:
Consider linking to other Node.js Section articles that are related to your article. Some mention testing, EJS, Node.js best practises and developing a Node.js app using Express.